include/fiction/algorithms/iter/gray_code_iterator.hpp

```diff
@@ -0,0 +1,295 @@
+//
+// Created by Jan Drewniok on 30.08.23.
+//
+
+#ifndef FICTION_GRAY_CODE_ITERATOR_HPP
+#define FICTION_GRAY_CODE_ITERATOR_HPP
+
+#include <cstdint>
+#include <iterator>
+#include <vector>
+
+namespace fiction
+{
+
+class gray_code_iterator
+{
+  public:
+    /**
+     * Constructs a Gray Code Iterator with specified range.
+     *
+     * This constructor creates a Gray Code Iterator that iterates over a range of decimal numbers
+     * and their corresponding Gray code representations. The iterator will start from the given
+     * `start` number and continue up to the `end` number (inclusive) in Gray code order.
+     *
+     * @param start The starting decimal number for the iterator.
+     * @param end The ending decimal number for the iterator.
+     */
+    explicit gray_code_iterator(const uint64_t start, const uint64_t end) noexcept :
```

Comment:
I wouldn't specify an end of the iteration here. Since you have implemented comparison operators against integers, you can use those in a loop to determine termination conditions. Would also make this code a little shorter even.

As per STL standards, ranges have `begin()` and `end()` functions that return iterators. All STL containers for instance. Iterators themselves should be comparable against each other.